### ISSUE TITLE
Allow specifying the environment of the mod (CurseForge only)

### ISSUE THREAD

**Issue by Steveplays28** [Kir-Antipov/mc-publish#64]
Since CurseForge has a checkbox for the environment (`server`/`client`) that is bound to the uploaded files (instead of the project, like Modrinth), every time a file is uploaded this checkbox needs to be set.
Is there any way to set this in the config that I'm missing?

**Comment by Kir-Antipov** [comment]
No, at the moment this is impossible. They hadn't have this feature when the last version of `mc-publish` was published. I hope to implement support for environments in v4.1 (it won't be CurseForge-only feature).

**Comment by Steveplays28** [comment]
Alright, thanks.
Just curious, how are you going to implement this for GitHub and Modrinth?

**Comment by Kir-Antipov** [comment]
As you pointed out, it is possible to specify the environments supported by your mod on Modrinth as well. The only difference is that CurseForge allows you to specify environments per version, while on Modrinth, it is a project-wide option. I believe it would be reasonable to assume that you would want your Modrinth project to reflect the capabilities of the latest version of your mod. Therefore, `mc-publish` could update the current settings of the project based on the version being published, if necessary.

As for GitHub, the `environment` input would remain unused, just as `github-loaders`, `github-dependencies`, and `github-game-versions`. There's a discussion on that (#60), so you may leave you suggestions and thoughts on how it could be implemented there.

**Comment by jh-devv** [comment]
+1

**Comment by UpcraftLP** [comment]
Any update on this?

It's been almost 2 years, and Curseforge has actually started using the environment flag in their app now: when exporting there is now an option to export only serverside mods.

![Image](https://github.com/user-attachments/assets/a5feaa1e-e82b-4a90-b23a-d0fd98b40282)

**Comment by Kira-NT** [comment]
![Yikes...](https://github.com/user-attachments/assets/adb28a8d-25bd-4169-8ff1-2c0c986d602b)

I guess I've properly GIMP3ed myself. It's not that this particular feature takes that long - it's that I've been slowly rewriting the entire project ~~in Rust~~, and now I'm at that stage where I can't really ship a new feature at the moment, 'cause the new codebase isn't there yet, and touching the old one would just open yet another can of worms. So, yeah, here we are :#

**Comment by Angelin01** [comment]
It's been even longer, and now Modrinth supports environment's per version too!

<img width="782" height="718" alt="Image" src="https://github.com/user-attachments/assets/f18583d9-2eff-4557-b29a-d31bc14cd18c" />

@Kira-NT I understand you've been trying to rewrite the project (for good reasons).

Do you need some help?
Additionally, if someone were to open a pull request adding support for this particular feature, are you wiling to merge it into the current version?

**Comment by Tiviacz1337** [comment]
Hey, still waiting for this feature, any updates about progress?

**Comment by Kira-NT** [comment]
> It's been even longer, and now Modrinth supports environment's per version too!

See, it's not me slacking off, I'm just strategically waiting for every supported platform to converge in terms of the supported feature set! :D

That's great news, actually. Now, if only Hangar also finally got a PATCH method for its `/versions` route, it would be perfect for v4. However, those folks somehow have an even worse development schedule than I do...

> Do you need some help?

Thanks for the consideration! All the help I need, unfortunately, lies beyond the realm of development. I probably already said that before, but the reason it takes so long is not because I'm really stuck implementing something, but because of real-life stuff. So I'm spending my quite limited time on things that I find both fun and important in the moment to work on, and that doesn't, erm, always mean mc-publish.

> Additionally, if someone were to open a pull request adding support for this particular feature, are you wiling to merge it into the current version?

It doesn't sit right with me to add features before fixing some little yet long-standing bugs the project has. Also, in the future I'd love to completely get rid of the JS origins of this tool, moving the legacy git history into a separate archived repo. That would mean I'd need to re-link the still-technically-runnable legacy versions of mc-publish with a backported v4 with a compatible API surface, so it would definitely help if the current one remains frozen :)

> Hey, still waiting for this feature, any updates about progress?

Can't really add anything beyond what I've already said. It's done when it's done, sowwy!

**Comment by rlnt** [comment]
Just as a quick note, the environment will be a required option in the future for CurseForge. This is likely a breaking change. They currently consider to set "server and client" as default, but if they decide against it, the action can no longer be used.

**Comment by UpcraftLP** [comment]
Update from CurseForge on this:

https://discord.com/channels/900128427150028811/1200093897053978644/1509157499809566851
<img width="1275" height="730" alt="Image" src="https://github.com/user-attachments/assets/7b4f50be-89b1-4db3-b223-41aec858fa48" />

**Comment by Kira-NT** [comment]
Breaking every tool that relies on your API instead of providing a sane default value is definitely *a* choice. But maybe it's for the better. Now I don't need to care about backward compatibility at all, since everyone will be forced to update, given that every single legacy version would be bricked by CurseForge anyways.

**Comment by UpcraftLP** [comment]
Any updates on this?

Less than 1 week left until the environment field becomes mandatory

**Comment by ThexXTURBOXx** [comment]
Should have commented here also, yeah...
@UpcraftLP Have you tried my fork? #157
Usage is analogous except for the `environments`: https://github.com/ThexXTURBOXx/BlockHelper/blob/4fdfae8bafbf2322a120b374966e3ebf7d79219d/.github/workflows/publish.yml#L117-L140

**Comment by Kira-NT** [comment]
> Any updates on this?

Thank you for your patience! This feature is now up and running. Unfortunately, had to dig into the old codebase, since the new one will not be ready for the foreseeable future. But oh well. Hopefully, this update will suffice until I'm able to finish everything currently sitting in my backlog.

If you were using floating tags like `v3` or `v3.3` (which now point to `v3.3.1`) instead of the fully specified `v3.3.0`, there's nothing you need to do for `mc-publish` to continue working *(although it's still a good idea to update the link to this action, since it has been moved to a new repo)* - it will automatically infer the correct environment for Fabric and Quilt mods, and for (Neo)Forge it will default to `client | server`, unless something else is explicitly specified.

Note that Modrinth has a pretty overengineered system of environments that covers all thinkable (and unthinkable, honestly) scenarios and use cases. Since it goes far beyond the simple "client"/"server"/"both" model expressible by mods, if you want to make use of it, you'll need to specify the correct environment yourself, either as a direct input to the action or as part of a custom `mc-publish`-branded block within your mod metadata. Here's an excerpt from the README describing its format:

<blockquote>

#### environment

A set of supported environments. By default, `mc-publish` attempts to infer this value from your project's metadata file (e.g., `fabric.mod.json`, `mods.toml`, `quilt.mod.json`, etc.).

```yaml
environment: client | server
```

Available environments (can be combined using the logical OR operator, as shown above):

- `client` - the project must be installed on the client.
- `client?` - the project may be installed on the client, but it is not required.
- `client*` - it is recommended to install the project on the client, but it is not required.
- `server` - the project must be installed on the server.
- `server?` - the project may be installed on the server, but it is not required.
- `server*` - it is recommended to install the project on the server, but it is not required.
- `singleplayer` - the project is supported in singleplayer, but not necessarily when connected to multiplayer servers.
- `dedicated-server` - the project is supported on dedicated servers, but not necessarily in singleplayer.

</blockquote>
